### PR TITLE
Import pogosh into u-root experimental

### PR DESCRIPTION
**Pull request by rjoleary**
This is a posix-compatible shell written in Go which is 20% implemented.

Import from: https://bitbucket.org/rjoleary/pogosh/

Signed-off-by: Ryan O'Leary <ryan@ryanoleary.ca>

### CODE REVIEW

File: pkg/pogosh/builtins.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+package shell
+
+import (
+	"strconv"
+)
+
+// DefaultBuiltins lists all the available builtins.
+func DefaultBuiltins() map[string]Builtin {
+	return map[string]Builtin{
+		"exit": BuiltinExit,
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/pogosh/builtins.go:14]: Yes, there are some gaping holes, but the over structure is sound. If we get this submitted, we can start involving other people to help fill them in.

File: pkg/pogosh/parser.go [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+
+// do_group          : 'do' compound_list 'done'
+//                   ;
+func parseDoGroup(s *State, t *tokenizer) {
+
+}
+
+// SimpleCommand ::= CmdPrefix SimpleCommand2 | CmdName CmdSuffix
+// SimpleCommand2 ::= CmdWord CmdSuffix |
+func parseSimpleCommand(s *State, t *tokenizer) command {
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/pogosh/parser.go:282]: It'd probably make more sense to write a separate lexer for the linux kernel cmdline. pogosh's lexer and parser apply special meaning to characters such as &&, ||, $, ...

File: pkg/pogosh/ast.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	cmd      command
+}
+
+// case word in cases esac
+type caseClause struct {
+	word  []byte
+	cases []caseItem
+}
+
+// pattern ) cmd
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/pogosh/ast.go:61]: The goal of pogosh is to implement posix shell spec. After reviewing the spec, it looks to be fairly attainable to implement (it's not the behemoth of bash or zsh).

[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/pogosh/ast.go:61]: I'm questioning the goal, not the implementation effort, though

[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/pogosh/ast.go:61]: There's no question that this is really good work, but I'd rather users write Go programs than shell scripts in u-root.

**Comment by hugelgupf** [comment]
> the original idea was the scripting language in u-root as the go compiler. That's not working out, sadly. This shell will give us a more general purpose command parser we can use in several places. So, while we don't really need it ... u-root was never intended to be just firmware ... this seems like something we can use.

Where would you like to use it? I've not seen too many places in u-root where scripting has been warranted, so I'm curious what you want this for. I'd still rather users write Go programs and compile them than script stuff.

**Pull request merged**